key-concepts/covenants.adoc

```diff
@@ -1,11 +1,174 @@
+}
+
+console.log(add(x, y))
+console.log(add(x, y))
+console.log(add(2, 3))
+console.log(add(x, y))
+console.log(x, y)
+----
+
+Outputs:
+[source]
+7
+17
+22
+47
+47 3
+
+|===
+
+These are contrived examples, but as you can see:
+
+- The pure function produces the same result each time, and `x` and `y`
+  are unmodified.
+
+- The non-pure function produces different results each time, even when
+  explicitly called with the initial inputs, and the value of `x` has
+  changed notably (a side effect of calling the function).
+
+
+== Covenants and blockchains
+
+An Interbit blockchain applies a single covenant. Therefore, the
+covenant fully controls how the blockchain's state can be modified.
+
+Other blockchains, such as the digital currencies BitCoin and Ethereum,
+use a single, public blockchain. This means that they have a single
+shared state that all participants must use. To provide flexibility,
+these blockchains permit arbitrary data storage on the blockchain.
+Unfortunately, this means that the state of these blockchains has become
+quite messy, and the storage requirements are surprising: in June 2018,
+BitCoin's blockchain size was 173 gigabytes, and in May 2018, Ethereum's
+blockchain size surpassed 1 terabyte).
+
+The Interbit philosophy is that each blockchain should represent a
+single concern. Instead of a single blockchain recording all of the
+personal details of a group of people, each person should record their
+personal details on their own blockchain, and the group should
+participate in a directory blockchain that records just the membership
+of each person in the group.
+
+This philosophy means that the covenant for an Interbit blockchain is
+often very simple. By itself, one blockchain with its covenant, performs
+a small set of tasks and maintains a small state. Simple covenants are
+easy to test. Confidence in the covenant, and therefore the blockchain,
+is high because it is easy to prove that it operates as designed.
+
+image:/architecture/topology/img/hub_and_spoke.svg["The Hub and Spoke
+molecule", 175, 194, role="right"]
+When several blockchains need to be used together, an administrative
+blockchain can be created, with a covenant that understands how those
+blockchains should inter-operate. The administrative covenant can also
+be simple, since it can be written knowing all of the actions that the
+blockchains support, and that they are going to behave as expected. This
+is the link:/architecture/topology/patterns.adoc#_hub_and_spoke[Hub and
+Spoke] topology.
+
+Eventually, a sophisticated application may manage many, many
+blockchains, performing complex operations that have been created almost
+entirely from the combined interaction of many simple blockchains. For
+example, if a small city powered all of its operations and services with
+Interbit blockchain technology, including blockchains for all of its
+citizens, service providers, policies, etc., there could be many
+millions of blockchains in use.
```

Comment:
We may want to also note that parts of the application can be shared between parties which allows the benefits of shared trust that blockchains facilitate, but other parts can be hosted on other nodes that remain private. (Or perhaps that's a different concept)